haskell-src/Concordium/Types/Queries.hs

```diff
@@ -1062,3 +1063,51 @@ data WinningBaker = WinningBaker
 --  record field is turned into the label @round@ in the corresponding JSON representation
 --  of the @PendingUpdate@.
 $(deriveJSON defaultOptions{fieldLabelModifier = firstLower . dropWhile isLower} ''WinningBaker)
+
+-- | A failing result of a dry run execution.
+data DryRunError
+    = -- | The current block state is undefined.
+      DryRunErrorNoState
+    | -- | The requested block was not found, so its state could not be loaded.
+      DryRunErrorBlockNotFound
+    | -- | The specified account was not found.
+      DryRunErrorAccountNotFound
+    | -- | The specified instance was not found.
+      DryRunErrorInstanceNotFound
+    | -- | The amount to mint would overflow the total CCD supply.
+      DryRunErrorAmountOverLimit {dreMaximumMintAmount :: !Amount}
+    | -- | The balance of the sender account is not sufficient to pay for the operation.
+      DryRunErrorBalanceInsufficient {dreRequiredAmount :: !Amount, dreAvailableAmount :: !Amount}
+    | -- | The energy supplied for the transaction is not sufficient to check the transaction
+      -- header.
+      DryRunErrorEnergyInsufficient {dreEnergyRequired :: !Energy}
+
+-- | A successful result of a dry run execution.
+-- These do not cover all successful results, just ones where the protobuf encoding cannot fail.
+data DryRunSuccess
+    = -- | The block state was loaded.
+      DryRunSuccessBlockStateLoaded
+        { -- | Current timestamp (taken from the block).
+          drsCurrentTimestamp :: !Timestamp,
+          -- | Block hash of the block the state was loaded from.
+          drsBlockHash :: !BlockHash,
+          -- | The protocol version determined by the block.
+          drsProtocolVersion :: !ProtocolVersion
+        }
+    | -- | The account info was successfully retrieved.
+      DryRunSuccessAccountInfo {drsAccountInfo :: !AccountInfo}
+    | -- | The smart contract instance info was successfully retrieved.
+      DryRunSuccessInstanceInfo {drsInstanceInfo :: !Wasm.InstanceInfo}
+    | -- | The current timestamp was successfully set.
+      DryRunSuccessTimestampSet
+    | -- | The requested amount was minted to the account.
+      DryRunSuccessMintedToAccount
```

Comment:
I don't think this would be a significant improvement.
- It is redundant, and there's not any real reason that the client should forget what it just asked for.
- Should the client be expected to deal with the situation where the response does not match the request?
- It opens the gate for parroting more information back to the client (e.g. for `BlockNotFound`, which block was not found), and there does not seem to be a clear line between when that would be useful and when it would not.